### ISSUE TITLE
CA1816 (Call GC.SuppressFinalize correctly) should not be reported when called inside IAsyncDisposable.DisposeAsync()

### ISSUE THREAD

**Issue by IIARROWS** [dotnet/roslyn-analyzers#4869]
### Analyzer

**Diagnostic ID**: [CA1816](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1816)

### Describe the improvement

This rule shouldn't be caused when `GC.SuppressFinalize` is called inside an implementation of `IAsyncDisposable.DisposeAsync` as it's part of the pattern as described here: https://docs.microsoft.com/dotnet/standard/garbage-collection/implementing-disposeasync

**Comment by Youssef1313** [comment]
This was supposed to be fixed with https://github.com/dotnet/roslyn-analyzers/pull/3679.

@IIARROWS Can you provide a minimal repro? and are you using the old deprecated FxCopAnalyzers, or the new NetAnalyzers package? and which version?

**Comment by IIARROWS** [comment]
I... don't know.
I'm using VS 2019 16.9.0 Preview 4, targeting .NET 5 in a Blazor WebAssembly project.
I don't have any idea how should I check what version I'm using.

**Comment by Youssef1313** [comment]
Using the following code from the article you're referring to, I don't get a warning from CA1816:

```csharp
using System;
using System.IO;
using System.Text.Json;
using System.Threading.Tasks;

public class ExampleAsyncDisposable : IAsyncDisposable, IDisposable
{
    private Utf8JsonWriter _jsonWriter = new Utf8JsonWriter(new MemoryStream());

    public void Dispose()
    {
        Dispose(disposing: true);
        GC.SuppressFinalize(this);
    }

    public async ValueTask DisposeAsync()
    {
        await DisposeAsyncCore();

        Dispose(disposing: false);
        GC.SuppressFinalize(this);
    }

    protected virtual void Dispose(bool disposing)
    {
        if (disposing)
        {
            _jsonWriter?.Dispose();
        }

        _jsonWriter = null;
    }

    protected virtual async ValueTask DisposeAsyncCore()
    {
        if (_jsonWriter is not null)
        {
            await _jsonWriter.DisposeAsync();
        }

        _jsonWriter = null;
    }
}
```

@IIARROWS Does the code above generate a warning for you? If it doesn't, can you post a code that generates the warning?

**Comment by Youssef1313** [comment]
@IIARROWS Ignore my above comment, I was using a preview version of the SDK which have this problem solved. You'll need to use the NuGet package instead of analyzers from the SDK until the fix flows into the SDK.

@mavasani What's the next version of SDK that will have the fix?

**Comment by mavasani** [comment]
> What's the next version of SDK that will have the fix?

It should be in .NET 6.0 SDK

**Comment by Uzivatel919** [comment]
Also [rule documentation](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1816) should be upgraded . Is there some schedule on this?

**Comment by bill-poole** [comment]
Is there an update on this? Just checking whether it has slipped through the cracks.

When I suppress CA1816 in source, I get an IDE0079 warning (remove unnecessary suppression). So I can choose between CA1816, IDE0079, or unnecessarily calling `GC.SuppressFinalize` in `DisposeAsync`.
